source/opt/value_number_table.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_VALUE_NUMBER_TABLE_H_
+#define LIBSPIRV_OPT_VALUE_NUMBER_TABLE_H_
+
+#include <cstdint>
+#include <unordered_map>
+#include "instruction.h"
+#include "ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+// Uses as the comparison in the value number table to determine if two
+// instruction compute the same value.
+class ComputeSameValue {
+ public:
+  bool operator()(const ir::Instruction& lhs, const ir::Instruction& rhs) const;
+};
+
+// The hash function used in the value number table.
+class ValueTableHash {
+ public:
+  std::size_t operator()(const spvtools::ir::Instruction& inst) const noexcept;
```

Comment:
The noexcept is causing the VS2013 build to fail.  However, I believe I need it for other builds.  I'll have to look into that.  I might need some macro magic to fix this.